datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -374,64 +346,130 @@ impl Stream for CrossJoinStream {
     fn poll_next_impl(
         &mut self,
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<Option<Result<RecordBatch>>> {
+        let schema = self.schema();
+
         let build_timer = self.join_metrics.build_time.timer();
         let (left_data, _) = match ready!(self.left_fut.get(cx)) {
             Ok(left_data) => left_data,
             Err(e) => return Poll::Ready(Some(Err(e))),
         };
         build_timer.done();
 
-        if left_data.num_rows() == 0 {
+        // If the left batch is empty, we can return `Poll::Ready(None)` immediately.
+        if left_data
+            .first()
+            .map_or(true, |first_batch| first_batch.num_rows() == 0)
+        {
             return Poll::Ready(None);
         }
 
-        if self.left_index > 0 && self.left_index < left_data.num_rows() {
-            let join_timer = self.join_metrics.join_time.timer();
-            let right_batch = {
-                let right_batch = self.right_batch.lock();
-                right_batch.clone().unwrap()
-            };
-            let result =
-                build_batch(self.left_index, &right_batch, left_data, &self.schema);
-            self.join_metrics.input_rows.add(right_batch.num_rows());
-            if let Ok(ref batch) = result {
-                join_timer.done();
-                self.join_metrics.output_batches.add(1);
-                self.join_metrics.output_rows.add(batch.num_rows());
+        match &self.curr_probe_batch {
+            Some(curr_probe_batch)
+                if self.right_row_index < curr_probe_batch.num_rows() =>
+            {
+                // Right batch has some unpaired rows, continue with the next row.
+                let result_batch = build_batch(
+                    &mut self.left_batch_index,
+                    &mut self.right_row_index,
+                    &mut self.join_metrics,
+                    left_data,
+                    curr_probe_batch,
+                    schema,
+                )?;
+                Poll::Ready(Some(Ok(result_batch)))
             }
-            self.left_index += 1;
-            return Poll::Ready(Some(result));
-        }
-        self.left_index = 0;
-        self.right
-            .poll_next_unpin(cx)
-            .map(|maybe_batch| match maybe_batch {
-                Some(Ok(batch)) => {
-                    let join_timer = self.join_metrics.join_time.timer();
-                    let result =
-                        build_batch(self.left_index, &batch, left_data, &self.schema);
+            // If it is the first poll or the right batch is fully processed:
+            _ => match ready!(self.right.poll_next_unpin(cx)) {
+                None => Poll::Ready(None),
+                Some(Ok(right_batch)) => {
+                    // New batch arrives, reset the indices.
+                    self.left_batch_index = 0;
+                    self.right_row_index = 0;
+                    let result_batch = build_batch(
```

Comment:
Why don't you make the build_batch a member method of CrossJoinStream? it seems suitable since it takes a lot of member attributes.